src/unfold.rs

```diff
@@ -0,0 +1,173 @@
+use crate::errors::Error;
+use crate::Github;
+use futures::stream;
+use hyperx::header::{Link, RelationType};
+use reqwest::Url;
+use serde::de::DeserializeOwned;
+
+pub async fn unfold<Source, StreamOk>(
+    github: Github,
+    initial: Result<(Option<Link>, Source), Error>,
+    to_items: Box<dyn Fn(Source) -> Vec<StreamOk> + Send + Sync>,
+) -> crate::Stream<StreamOk>
+where
+    StreamOk: 'static + Send + Sync,
+    Source: DeserializeOwned + 'static + Send + Sync,
+{
+    let state = StreamState::new(github, initial, to_items);
+
+    Box::pin(stream::unfold(state, move |state| async {
+        state.next().await
+    }))
+}
+
+struct StreamState<Source, StreamOk>
+where
+    StreamOk: 'static + Send + Sync,
+    Source: DeserializeOwned + 'static + Send + Sync,
+{
+    github: Github,
+    items: Option<Result<Vec<StreamOk>, Error>>,
+    to_items: Box<dyn Fn(Source) -> Vec<StreamOk> + Send + Sync>,
+    next_page: Option<Link>,
+}
+
+impl<Source, StreamOk> StreamState<Source, StreamOk>
+where
+    StreamOk: 'static + Send + Sync,
+    Source: DeserializeOwned + 'static + Send + Sync,
+{
+    fn new(
+        github: Github,
+        initial: Result<(Option<Link>, Source), Error>,
+        to_items: Box<dyn Fn(Source) -> Vec<StreamOk> + Send + Sync>,
+    ) -> Self {
+        let dummy = Self {
+            github,
+            to_items,
+            items: None,
+            next_page: None,
+        };
+
+        dummy.load_state(initial)
+    }
+
+    async fn next(self) -> Option<(Result<StreamOk, Error>, Self)> {
+        let mut state = self;
+        loop {
+            match state.flat_next() {
+                FetcherState::NextReady(item, state) => {
+                    return Some((item, state));
+                }
+                FetcherState::Empty => {
+                    return None;
+                }
+                FetcherState::FetchNextPage(next_state) => {
+                    state = next_state.fetch_next_page().await;
+                }
+            }
+        }
+    }
+
+    fn flat_next(mut self) -> FetcherState<StreamOk, Self> {
+        if let Some(resultitems) = self.items.take() {
+            match resultitems {
+                Ok(mut items) => {
+                    if let Some(item) = items.pop() {
```

Comment:
The use of `items.pop()` without reversing `items` results in wrong order. `[5, 4, 3, 2, 1, 10, 9, 8, 7, 6]`